# Gendering the talks 

One of our areas of interest are to see if TED talks by men and women are different. If there are differences, we would like to see if we can detect the gender of a speaker based on the words in a TED talk transcript. To pursue these avenues, we begin by "gendering" talks when possible. 

In this notebook, we use the genders of the speakers to "gender" the TED talks themselves. A talk with one speaker inherits the gender of the speaker. For a talk with two speakers, if the genders of the speakers are the same, then we proceed as if the talk had one speaker. For talks with two speakers where the genders of the speakers is not the same, we place these talks to the side. 

In [2]:
import pandas as pd
import csv
import string

In [3]:
# Load the gendered speaker file:
speakers = pd.read_csv("speakers_with_gender.csv")

In [4]:
ted_only = pd.read_csv('../data/Release_v0/TEDonly_final.csv')
ted_plus = pd.read_csv('../data/Release_v0/TEDplus_final.csv')

In [5]:
ted_only.head()

Unnamed: 0  Unnamed: 0.1  Talk_ID  \
0           0             0        1   
1           1             1        7   
2           2             2       53   
3           3             3       66   
4           4             4       92   

                                          public_url  \
0  https://www.ted.com/talks/al_gore_on_averting_...   
1  https://www.ted.com/talks/david_pogue_says_sim...   
2  https://www.ted.com/talks/majora_carter_s_tale...   
3  https://www.ted.com/talks/ken_robinson_says_sc...   
4  https://www.ted.com/talks/hans_rosling_shows_t...   

                          headline  \
0      Averting the climate crisis   
1                 Simplicity sells   
2              Greening the ghetto   
3      Do schools kill creativity?   
4  The best stats you've ever seen   

                                         description    event duration  \
0  With the same humor and humanity he exuded in ...  TED2006  0:16:17   
1  New York Times columnist David Pogue takes aim...  TED2006  0:21:26   
2  In an emotionally charged talk, MacArthur-winn...  TED2006  0:18:36   
3  Sir Ken Robinson makes an entertaining and pro...  TED2006  0:19:24   
4  You've never seen data presented like this. Wi...  TED2006  0:19:50   

  published                                               tags     views  \
0   6/27/06  alternative energy,cars,global issues,climate ...   3266733   
1   6/27/06  simplicity,entertainment,interface design,soft...   1702201   
2   6/27/06  MacArthur grant,cities,green,activism,politics...   2000421   
3   6/27/06  children,teaching,creativity,parenting,culture...  51614087   
4   6/27/06  demo,Asia,global issues,visualizations,global ...  12662135   

                                                text      speaker_1 speaker_2  \
0    Thank you so much, Chris. And it's truly a g...        Al Gore       NaN   
1    (Music: "The Sound of Silence," Simon & Garf...    David Pogue       NaN   
2    If you're here today — and I'm very happy th...  Majora Carter       NaN   
3    Good morning. How are you?    (Laughter)    ...   Ken Robinson       NaN   
4    About 10 years ago, I took on the task to te...   Hans Rosling       NaN   

  speaker_3 speaker_4  
0       NaN       NaN  
1       NaN       NaN  
2       NaN       NaN  
3       NaN       NaN  
4       NaN       NaN

In [6]:
# Set the talk ID as the index and drop the unnecessary first two columns: 
ted_only = ted_only.set_index('Talk_ID')
ted_only = ted_only.drop(columns = ['Unnamed: 0', 'Unnamed: 0.1'])

In [7]:
ted_only.head()

public_url  \
Talk_ID                                                      
1        https://www.ted.com/talks/al_gore_on_averting_...   
7        https://www.ted.com/talks/david_pogue_says_sim...   
53       https://www.ted.com/talks/majora_carter_s_tale...   
66       https://www.ted.com/talks/ken_robinson_says_sc...   
92       https://www.ted.com/talks/hans_rosling_shows_t...   

                                headline  \
Talk_ID                                    
1            Averting the climate crisis   
7                       Simplicity sells   
53                   Greening the ghetto   
66           Do schools kill creativity?   
92       The best stats you've ever seen   

                                               description    event duration  \
Talk_ID                                                                        
1        With the same humor and humanity he exuded in ...  TED2006  0:16:17   
7        New York Times columnist David Pogue takes aim...  TED2006  0:21:26   
53       In an emotionally charged talk, MacArthur-winn...  TED2006  0:18:36   
66       Sir Ken Robinson makes an entertaining and pro...  TED2006  0:19:24   
92       You've never seen data presented like this. Wi...  TED2006  0:19:50   

        published                                               tags  \
Talk_ID                                                                
1         6/27/06  alternative energy,cars,global issues,climate ...   
7         6/27/06  simplicity,entertainment,interface design,soft...   
53        6/27/06  MacArthur grant,cities,green,activism,politics...   
66        6/27/06  children,teaching,creativity,parenting,culture...   
92        6/27/06  demo,Asia,global issues,visualizations,global ...   

            views                                               text  \
Talk_ID                                                                
1         3266733    Thank you so much, Chris. And it's truly a g...   
7         1702201    (Music: "The Sound of Silence," Simon & Garf...   
53        2000421    If you're here today — and I'm very happy th...   
66       51614087    Good morning. How are you?    (Laughter)    ...   
92       12662135    About 10 years ago, I took on the task to te...   

             speaker_1 speaker_2 speaker_3 speaker_4  
Talk_ID                                               
1              Al Gore       NaN       NaN       NaN  
7          David Pogue       NaN       NaN       NaN  
53       Majora Carter       NaN       NaN       NaN  
66        Ken Robinson       NaN       NaN       NaN  
92        Hans Rosling       NaN       NaN       NaN

First, we want to know which talks only have one speaker. To do this, we only select the rows that `NaN` for the second speaker. 

In [8]:
talk1speak = ted_only[ted_only['speaker_2'].isnull()]

# https://stackoverflow.com/questions/43831539/how-to-select-rows-with-nan-in-particular-column

In [9]:
talk1speak.shape

(966, 13)

We have 966 talks that have only one speaker. 


To figure out which talks have exactly two speakers, we select the rows that have a second speaker, but **not** a third one:  

In [10]:
s2temp = ted_only[ted_only['speaker_3'].isnull()]
talk2speak = s2temp[~s2temp['speaker_2'].isnull()]

In [11]:
talk2speak.shape

(25, 13)

We have 25 talks that have two speakers. 


To figure out which talks have three speakers, we select the rows that have a third speaker, but **not** a fourth one:  

In [12]:
s3temp = ted_only[ted_only['speaker_4'].isnull()]
talk3speak = s3temp[~s3temp['speaker_3'].isnull()]

In [13]:
talk3speak.shape

(0, 13)

There are no talks with exactly three speakers. 


To figure out which talks have four speakers, we select the rows that have a fourth speaker:  

In [14]:
talk4speak = ted_only[~ted_only['speaker_4'].isnull()]


In [15]:
talk4speak.shape

(1, 13)

As a quick check, we check if the total number of rows in the `talkNspeak` data frames equal the number of rows in `ted_only`:

In [16]:
ted_only.shape[0] == talk1speak.shape[0] + talk2speak.shape[0] + talk3speak.shape[0] + talk4speak.shape[0]

True

## Adding gender to the talks

Now that we have separate the talks into smaller data frames, each representing the number of speakers, we will be working to "gender" the talks. 

In [17]:
gender_slice = speakers[["speaker","Gender_handcheck"]]

In [18]:
gender_slice

speaker     Gender_handcheck
0               Al Gore                 male
1           David Pogue                 male
2         Majora Carter               female
3          Ken Robinson                 male
4          Hans Rosling                 male
..                  ...                  ...
865   Improv Everywhere   they (group mixed)
866     Quixotic Fusion   they (group mixed)
867  Silk Road Ensemble   they (group mixed)
868             The LXD   they (group mixed)
869  Rodrigo y Gabriela  they (male, female)

[870 rows x 2 columns]

In [19]:
# We want to take speaker gender from speakers and put it into the talks

talk1speak = talk1speak.merge(gender_slice, left_on = "speaker_1", right_on = "speaker", how = "left")


In [20]:
talk1speak

public_url  \
0    https://www.ted.com/talks/al_gore_on_averting_...   
1    https://www.ted.com/talks/david_pogue_says_sim...   
2    https://www.ted.com/talks/majora_carter_s_tale...   
3    https://www.ted.com/talks/ken_robinson_says_sc...   
4    https://www.ted.com/talks/hans_rosling_shows_t...   
..                                                 ...   
963  https://www.ted.com/talks/atul_gawande_want_to...   
964  https://www.ted.com/talks/jacob_collier_a_one_...   
965  https://www.ted.com/talks/anna_rosling_ronnlun...   
966  https://www.ted.com/talks/guy_winch_how_to_fix...   
967  https://www.ted.com/talks/chuck_nice_a_funny_l...   

                                              headline  \
0                          Averting the climate crisis   
1                                     Simplicity sells   
2                                  Greening the ghetto   
3                          Do schools kill creativity?   
4                      The best stats you've ever seen   
..                                                 ...   
963        Want to get great at something? Get a coach   
964                       A one-man musical phenomenon   
965  See how the rest of the world lives, organized...   
966                          How to fix a broken heart   
967  A funny look at the unintended consequences of...   

                                           description    event duration  \
0    With the same humor and humanity he exuded in ...  TED2006  0:16:17   
1    New York Times columnist David Pogue takes aim...  TED2006  0:21:26   
2    In an emotionally charged talk, MacArthur-winn...  TED2006  0:18:36   
3    Sir Ken Robinson makes an entertaining and pro...  TED2006  0:19:24   
4    You've never seen data presented like this. Wi...  TED2006  0:19:50   
..                                                 ...      ...      ...   
963  How do we improve in the face of complexity? A...  TED2017  0:16:58   
964  Jacob Collier is a one-man band and force of n...  TED2017  0:15:14   
965  What does it look like when someone in Sweden ...  TED2017  0:11:41   
966  At some point in our lives, almost every one o...  TED2017  0:12:25   
967  Technology should work for us, but what happen...  TED2017  0:10:42   

    published                                               tags     views  \
0     6/27/06  alternative energy,cars,global issues,climate ...   3266733   
1     6/27/06  simplicity,entertainment,interface design,soft...   1702201   
2     6/27/06  MacArthur grant,cities,green,activism,politics...   2000421   
3     6/27/06  children,teaching,creativity,parenting,culture...  51614087   
4     6/27/06  demo,Asia,global issues,visualizations,global ...  12662135   
..        ...                                                ...       ...   
963  12/15/17  Surgery,goal-setting,potential,teaching,life,s...   1493621   
964   1/12/18  visualizations,performance art,live music,voca...    469560   
965   1/18/18  global issues,visualizations,photography,data,...   1330850   
966    2/5/18  love,addiction,depression,compassion,personal ...   2315815   
967   2/27/18  AI,funny,comedy,computers,technology,social me...    541730   

                                                  text              speaker_1  \
0      Thank you so much, Chris. And it's truly a g...                Al Gore   
1      (Music: "The Sound of Silence," Simon & Garf...            David Pogue   
2      If you're here today — and I'm very happy th...          Majora Carter   
3      Good morning. How are you?    (Laughter)    ...           Ken Robinson   
4      About 10 years ago, I took on the task to te...           Hans Rosling   
..                                                 ...                    ...   
963    I don't come to you today as an expert. I co...           Atul Gawande   
964    (Music)    (Singing) I was walking down the ...          Jacob Collier   
965    What images do we see from the rest of the w...  Anna Rosling Rö

In [21]:
# Remove the second speaker column and rename the last column to "talk_gender"

talk1speak.drop(columns = ['speaker'])
talk1speak.rename(columns={"Gender_handcheck": "talk_gender"})

public_url  \
0    https://www.ted.com/talks/al_gore_on_averting_...   
1    https://www.ted.com/talks/david_pogue_says_sim...   
2    https://www.ted.com/talks/majora_carter_s_tale...   
3    https://www.ted.com/talks/ken_robinson_says_sc...   
4    https://www.ted.com/talks/hans_rosling_shows_t...   
..                                                 ...   
963  https://www.ted.com/talks/atul_gawande_want_to...   
964  https://www.ted.com/talks/jacob_collier_a_one_...   
965  https://www.ted.com/talks/anna_rosling_ronnlun...   
966  https://www.ted.com/talks/guy_winch_how_to_fix...   
967  https://www.ted.com/talks/chuck_nice_a_funny_l...   

                                              headline  \
0                          Averting the climate crisis   
1                                     Simplicity sells   
2                                  Greening the ghetto   
3                          Do schools kill creativity?   
4                      The best stats you've ever seen   
..                                                 ...   
963        Want to get great at something? Get a coach   
964                       A one-man musical phenomenon   
965  See how the rest of the world lives, organized...   
966                          How to fix a broken heart   
967  A funny look at the unintended consequences of...   

                                           description    event duration  \
0    With the same humor and humanity he exuded in ...  TED2006  0:16:17   
1    New York Times columnist David Pogue takes aim...  TED2006  0:21:26   
2    In an emotionally charged talk, MacArthur-winn...  TED2006  0:18:36   
3    Sir Ken Robinson makes an entertaining and pro...  TED2006  0:19:24   
4    You've never seen data presented like this. Wi...  TED2006  0:19:50   
..                                                 ...      ...      ...   
963  How do we improve in the face of complexity? A...  TED2017  0:16:58   
964  Jacob Collier is a one-man band and force of n...  TED2017  0:15:14   
965  What does it look like when someone in Sweden ...  TED2017  0:11:41   
966  At some point in our lives, almost every one o...  TED2017  0:12:25   
967  Technology should work for us, but what happen...  TED2017  0:10:42   

    published                                               tags     views  \
0     6/27/06  alternative energy,cars,global issues,climate ...   3266733   
1     6/27/06  simplicity,entertainment,interface design,soft...   1702201   
2     6/27/06  MacArthur grant,cities,green,activism,politics...   2000421   
3     6/27/06  children,teaching,creativity,parenting,culture...  51614087   
4     6/27/06  demo,Asia,global issues,visualizations,global ...  12662135   
..        ...                                                ...       ...   
963  12/15/17  Surgery,goal-setting,potential,teaching,life,s...   1493621   
964   1/12/18  visualizations,performance art,live music,voca...    469560   
965   1/18/18  global issues,visualizations,photography,data,...   1330850   
966    2/5/18  love,addiction,depression,compassion,personal ...   2315815   
967   2/27/18  AI,funny,comedy,computers,technology,social me...    541730   

                                                  text              speaker_1  \
0      Thank you so much, Chris. And it's truly a g...                Al Gore   
1      (Music: "The Sound of Silence," Simon & Garf...            David Pogue   
2      If you're here today — and I'm very happy th...          Majora Carter   
3      Good morning. How are you?    (Laughter)    ...           Ken Robinson   
4      About 10 years ago, I took on the task to te...           Hans Rosling   
..                                                 ...                    ...   
963    I don't come to you today as an expert. I co...           Atul Gawande   
964    (Music)    (Singing) I was walking down the ...          Jacob Collier   
965    What images do we see from the rest of the w...  Anna Rosling Rö

In [22]:
talk1speak.shape

(968, 15)

Two extra rows were added during this process. This is likely due to speakers having the same name as noted in "TED talks as Data" paper. 

We need to remove these extra rows:

In [26]:
talk1_dup = talk1speak[talk1speak.duplicated(['headline'])]


In [27]:
talk1_dup


public_url  \
46   https://www.ted.com/talks/neil_gershenfeld_on_...   
582  https://www.ted.com/talks/amanda_palmer_the_ar...   

                                 headline  \
46   Unleash your creativity in a Fab Lab   
582                     The art of asking   

                                           description    event duration  \
46   MIT professor Neil Gershenfeld talks about his...  TED2006  0:17:18   
582  Don't make people pay for music, says Amanda P...  TED2013  0:13:47   

    published                                               tags    views  \
46    2/19/07  code,engineering,materials,computers,science,i...   744087   
582    3/1/13       entertainment,performance art,business,music  9417001   

                                                  text         speaker_1  \
46     This meeting has really been about a digital...  Neil Gershenfeld   
582    (Breathes in)    (Breathes out)    So, I did...     Amanda Palmer   

    speaker_2 speaker_3 speaker_4           speaker Gender_handcheck  
46        NaN       NaN       NaN  Neil Gershenfeld             male  
582       NaN       NaN       NaN     Amanda Palmer           female

In [30]:
talk1speak[talk1speak["speaker_1"] == "Neil Gershenfeld"]

public_url  \
45  https://www.ted.com/talks/neil_gershenfeld_on_...   
46  https://www.ted.com/talks/neil_gershenfeld_on_...   

                                headline  \
45  Unleash your creativity in a Fab Lab   
46  Unleash your creativity in a Fab Lab   

                                          description    event duration  \
45  MIT professor Neil Gershenfeld talks about his...  TED2006  0:17:18   
46  MIT professor Neil Gershenfeld talks about his...  TED2006  0:17:18   

   published                                               tags   views  \
45   2/19/07  code,engineering,materials,computers,science,i...  744087   
46   2/19/07  code,engineering,materials,computers,science,i...  744087   

                                                 text         speaker_1  \
45    This meeting has really been about a digital...  Neil Gershenfeld   
46    This meeting has really been about a digital...  Neil Gershenfeld   

   speaker_2 speaker_3 speaker_4           speaker Gender_handcheck  
45       NaN       NaN       NaN  Neil Gershenfeld             male  
46       NaN       NaN       NaN  Neil Gershenfeld             male

In [31]:
talk1speak[talk1speak["speaker_1"] == "Amanda Palmer"]

public_url           headline  \
581  https://www.ted.com/talks/amanda_palmer_the_ar...  The art of asking   
582  https://www.ted.com/talks/amanda_palmer_the_ar...  The art of asking   

                                           description    event duration  \
581  Don't make people pay for music, says Amanda P...  TED2013  0:13:47   
582  Don't make people pay for music, says Amanda P...  TED2013  0:13:47   

    published                                          tags    views  \
581    3/1/13  entertainment,performance art,business,music  9417001   
582    3/1/13  entertainment,performance art,business,music  9417001   

                                                  text      speaker_1  \
581    (Breathes in)    (Breathes out)    So, I did...  Amanda Palmer   
582    (Breathes in)    (Breathes out)    So, I did...  Amanda Palmer   

    speaker_2 speaker_3 speaker_4        speaker Gender_handcheck  
581       NaN       NaN       NaN  Amanda Palmer           female  
582       NaN       NaN       NaN  Amanda Palmer           female

In [33]:
speakers[speakers["speaker"] == "Amanda Palmer"]

speaker         occupation  \
441  Amanda Palmer  Musician, blogger   
665  Amanda Palmer           musician   

                                          introduction  \
441  Alt-rock icon Amanda Fucking Palmer believes w...   
665  Alt-rock icon Amanda Fucking Palmer believes w...   

                                               profile Gender_auto  MaleScore  \
441  Why you should listen\nAmanda Palmer commands ...      female        0.0   
665  Why you should listen\nAmanda Palmer commands ...      female        0.0   

     FemaleScore  NonBinaryScore Gender_handcheck  
441         15.0             4.0           female  
665         15.0             3.0           female

In [34]:
speakers[speakers["speaker"] == "Neil Gershenfeld"]

speaker                       occupation  \
757  Neil Gershenfeld  Physicist, personal fab pioneer   
803  Neil Gershenfeld                        Physicist   

                                          introduction  \
757  As Director of MIT’s Center for Bits and Atoms...   
803  As Director of MIT’s Center for Bits and Atoms...   

                                               profile Gender_auto  MaleScore  \
757  Why you should listen\nMIT's Neil Gershenfeld ...  undetected        2.0   
803  Why you should listen\nMIT's Neil Gershenfeld ...  undetected        2.0   

     FemaleScore  NonBinaryScore Gender_handcheck  
757          0.0             2.0             male  
803          0.0             2.0             male

Interestingly, it's that we have two copies of the same person in the speakers dataset. This means that we can drop either copy of the repeated rows in `talk1speak`:

In [35]:
talk1speak = talk1speak.drop_duplicates()
talk1speak.shape

(966, 15)

Now with that talks with just one speaker each have been "gendered," we turn our attention to the talks with multiple speakers. We start with those that have two speakers: 

In [36]:
talk2speak

public_url  \
Talk_ID                                                      
118      https://www.ted.com/talks/sergey_brin_and_larr...   
222      https://www.ted.com/talks/the_jill_and_julia_show   
224      https://www.ted.com/talks/roy_gould_and_curtis...   
246      https://www.ted.com/talks/tod_machover_and_dan...   
322      https://www.ted.com/talks/bruno_bowden_folds_w...   
385         https://www.ted.com/talks/toys_from_the_future   
481      https://www.ted.com/talks/pattie_maes_demos_th...   
881      https://www.ted.com/talks/debate_does_the_worl...   
988      https://www.ted.com/talks/david_byrne_sings_no...   
1156     https://www.ted.com/talks/robert_gupta_and_jos...   
1304     https://www.ted.com/talks/homaro_cantu_ben_roc...   
1722     https://www.ted.com/talks/two_young_scientists...   
1823     https://www.ted.com/talks/eric_berlow_and_sean...   
1948     https://www.ted.com/talks/steven_pinker_and_re...   
1964     https://www.ted.com/talks/bill_and_melinda_gat...   
1972     https://www.ted.com/talks/gabby_giffords_and_m...   
2300     https://www.ted.com/talks/alec_soth_stacey_bak...   
2611     https://www.ted.com/talks/silk_road_ensemble_r...   
2481     https://www.ted.com/talks/amanda_palmer_jherek...   
2770     https://www.ted.com/talks/serena_williams_gayl...   
2788     https://www.ted.com/talks/t_morgan_dixon_and_v...   
2851     https://www.ted.com/talks/jon_boogz_and_lil_bu...   
3625     https://www.ted.com/talks/shonda_rhimes_and_cy...   
3633     https://www.ted.com/talks/sebastian_thrun_and_...   
8420     https://www.ted.com/talks/stewart_brand_and_ch...   

                                                  headline  \
Talk_ID                                                      
118                                  The genesis of Google   
222                                The Jill and Julia Show   
224                   A preview of the WorldWide Telescope   
246            Inventing instruments that unlock new music   
322                            Blindfold origami and cello   
385                     Toys and materials from the future   
481                        Meet the SixthSense interaction   
881            Debate: Does the world need nuclear energy?   
988            "(Nothing But) Flowers" with string quartet   
1156                    On violin and cello, "Passacaglia"   
1304                                    Cooking as alchemy   
1722     Two young scientists break down plastics with ...   
1823                         Mapping ideas worth spreading   
1948                              The long reach of reason   
1964     Why giving away our wealth has been the most s...   
1972           Be passionate. Be courageous. Be your best.   
2300                 This is what enduring love looks like   
2611                           "St. James Infirmary Blues"   
2481                                        "Space Oddity"   
2770                        On tennis, love and motherhood   
2788                  When Black women walk, things change   
2851                         A dance to honor Mother Earth   
3625                            The future of storytelling   
3633                               What AI is -- and isn't   
8420     Mammoths resurrected, geoengineering and other...   

                                               description    event duration  \
Talk_ID                                                                        
118      Google co-founders Larry Page and Sergey Brin ...  TED2004  0:20:33   
222      Two TED favorites, Jill Sobule and Julia Sween...  TED2007  0:06:14   
224      Educator Roy Gould and researcher Curtis Wong ...  TED2008  0:06:42   
246      Tod Machover of MIT's Media Lab is devoted to ...  TED2008  0:20:41   
322      After Robert Lang's talk on origami at TED2008...  TED2008  0:02:58   
385      The Inventables guys, Zach Kaplan and Keith Sc...  TED2005  0:15:46   
481      This demo -- from Pattie Maes' lab at MIT, spe...